pkg/k8s/watchers/cilium_cidr_group.go

```diff
@@ -0,0 +1,227 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package watchers
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	cilium_v2_alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+func (k *K8sWatcher) onUpsertCIDRGroup(
+	cidrGroup *cilium_v2_alpha1.CiliumCIDRGroup,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+) error {
+	oldCidrGroup, ok := cidrGroupCache[cidrGroup.Name]
+	if ok && oldCidrGroup.Spec.DeepEqual(&cidrGroup.Spec) {
+		return nil
+	}
+
+	cidrGroupCpy := cidrGroup.DeepCopy()
+	cidrGroupCache[cidrGroup.Name] = cidrGroupCpy
+
+	err := k.updateCIDRGroupRefPolicies(cidrGroup.Name, cidrGroupCache, cnpCache, cs)
+
+	return err
+}
+
+func (k *K8sWatcher) onDeleteCIDRGroup(
+	cidrGroupName string,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+) error {
+	delete(cidrGroupCache, cidrGroupName)
+
+	err := k.updateCIDRGroupRefPolicies(cidrGroupName, cidrGroupCache, cnpCache, cs)
+
+	return err
+}
+
+func (k *K8sWatcher) updateCIDRGroupRefPolicies(
+	cidrGroup string,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+) error {
+	var errs []error
+	for key, cnp := range cnpCache {
```

Comment:
There was some discussion with Joe and I about making a separate refactoring commit which is clearly titled as "Unify CNP and CCNP handling in watchers" which would've made this very clear and have the space to explain why it's necessary. @pippolo84 if this is not too much work, it might be good to have this.

Reply:
Yep, np, I'm gonna split the commit.